src/ast/passes/codegen_llvm.cpp

```diff
@@ -3860,6 +3924,142 @@ void CodegenLLVM::createIncDec(Unop &unop)
+
+  b_.CreateStore(callback->getArg(1), nr_stack_frames);
+  b_.CreateStore(callback->getArg(2), seed_addr);
+
+  // uint64_t id = seed ^ (len * m);
+  b_.CreateStore(
+      b_.CreateXor(b_.CreateLoad(b_.getInt64Ty(), seed_addr),
+                   b_.CreateMul(b_.CreateIntCast(b_.CreateLoad(b_.getInt8Ty(),
+                                                               nr_stack_frames),
+                                                 b_.getInt64Ty(),
+                                                 false),
+                                m)),
+      id);
+
+  // int i = 0;
+  b_.CreateStore(b_.getInt8(0), i);
+
+  Function *parent = b_.GetInsertBlock()->getParent();
+  BasicBlock *while_cond = BasicBlock::Create(module_->getContext(),
+                                              "while_cond",
+                                              parent);
+  BasicBlock *while_body = BasicBlock::Create(module_->getContext(),
+                                              "while_body",
+                                              parent);
+  BasicBlock *while_end = BasicBlock::Create(module_->getContext(),
+                                             "while_end",
+                                             parent);
+  b_.CreateBr(while_cond);
+  b_.SetInsertPoint(while_cond);
+  auto *cond = b_.CreateICmp(CmpInst::ICMP_ULT,
+                             b_.CreateLoad(b_.getInt8Ty(), i),
+                             b_.CreateLoad(b_.getInt8Ty(), nr_stack_frames),
+                             "length.cmp");
+  b_.CreateCondBr(cond, while_body, while_end);
+
+  b_.SetInsertPoint(while_body);
+
+  // uint64_t k = stack[i];
+  Value *stack_ptr = b_.CreateGEP(b_.getInt64Ty(),
+                                  stack_addr,
+                                  b_.CreateLoad(b_.getInt8Ty(), i));
+  b_.CreateStore(b_.CreateLoad(b_.getInt64Ty(), stack_ptr), k);
+
+  // k *= m;
+  b_.CreateStore(b_.CreateMul(b_.CreateLoad(b_.getInt64Ty(), k), m), k);
+
+  // // k ^= k >> r
+  b_.CreateStore(b_.CreateXor(b_.CreateLoad(b_.getInt64Ty(), k),
+                              b_.CreateLShr(b_.CreateLoad(b_.getInt64Ty(), k),
+                                            r)),
+                 k);
+
+  // // k *= m
+  b_.CreateStore(b_.CreateMul(b_.CreateLoad(b_.getInt64Ty(), k), m), k);
+
+  // id ^= k
+  b_.CreateStore(b_.CreateXor(b_.CreateLoad(b_.getInt64Ty(), id),
+                              b_.CreateLoad(b_.getInt64Ty(), k)),
+                 id);
+
+  // id *= m
+  b_.CreateStore(b_.CreateMul(b_.CreateLoad(b_.getInt64Ty(), id), m), id);
+
+  // ++i
+  b_.CreateStore(b_.CreateAdd(b_.CreateLoad(b_.getInt8Ty(), i), b_.getInt8(1)),
+                 i);
+
+  b_.CreateBr(while_cond);
+  b_.SetInsertPoint(while_end);
+
+  b_.CreateLifetimeEnd(nr_stack_frames);
+  b_.CreateLifetimeEnd(seed_addr);
+  b_.CreateLifetimeEnd(i);
+  b_.CreateLifetimeEnd(k);
+
+  b_.CreateRet(b_.CreateLoad(b_.getInt64Ty(), id));
```

Comment:
I think we'll also need a `CreateLifetimeEnd(id)` between the `Load` and the `Ret`

Reply:
These lifetimes are a real pain (and aren't being done correctly in a lot of the rest of Codegen).

It'd be good if we could create a post-codegen pass to insert them automatically somehow. Why doesn't LLVM do this already though...?

Reply:
is this something that scoped deleters could help with?
could make it so that `CreateAllocaBPF` returns a tuple, containing both the pointer-to-buffer _and_ a ScopedExprDeleter, which would emits a `CreateLifetimeEnd()` when it leaves scope.  
so by default we'd always be cleaning up resources, and if we wanted to extend the lifetime we'd have to explicitly disarm the deleter.

Reply:
Yeah lexically scoped deleters makes a lot of sense to me.